### DISCUSSION TITLE
Is converting javascript to php possible?

### DISCUSSION THREAD

**Discussion by plaffy-guy** [rectorphp/rector#7018] [Q&A]
I am working on a project to where I have to convert valid javascript (es7) code to php code. Would it be possibly in theory to write rules such that I could find javascript syntax and then have it run through the transpiler and have it come out as php code? I wouldn't expect it to work out of the box I'm more just looking for whether or not it would be possible at all. 

Apologies if this is a dumb question.

**Comment by TomasVotruba** [comment]
Hi, great question. Of course it's possible :). All you need is AST-parser for Javascript. Then you can parser the code and map it to PHP nodes. Then you can compose them together and print with php-parser Standard printer. 

I've heard about Babel https://babeljs.io/, that might help you to start things of.

**Reply by plaffy-guy** [reply]
Ok sweet I was really hoping that would be the case. 

So to make sure that I'm understanding correctly, I could take babel and generate an AST from the input javacsript file. I would then need to map that AST to the format that php-parser uses so that it could run through rector. 

Is that on the right track? I'm still getting familiar with cross language transpiling and AST's in general so want to make I'm wrapping my head around it in the right direction at least.

**Reply by TomasVotruba** [reply]
> I would then need to map that AST to the format that php-parser uses so that it could run through rector.

This is one option, more clear and generics and uses php-parser know-how. Better for PHP dev.

Or you can go the lazy way and just write own string printer for Babel. Better for Javascript dev.

**Reply by plaffy-guy** [reply]
Perfect that makes total sense. The project that I'm working on will be ultimately to provide assistance and ease of use to a php dev so I'll go the route of reformatting the babel output to be in the php-parser format since that is what they would be familiar with. 

I really appreciate the help and responses!
